rllib/agents/dqn/apex.py

```diff
@@ -283,6 +374,256 @@ def add_apex_metrics(result: dict) -> dict:
+        self, num_samples_collected: Dict[ActorHandle, int]
+    ) -> None:
+        """Get samples from the replay buffer and place them on the learner queue.
+
+        Args:
+            num_samples_collected: A mapping from ActorHandle (RolloutWorker) to
+                number of samples returned by the remote worker. This is used to
+                implement training intensity which is the concept of triggering a
+                certain amount of training based on the number of samples that have
+                been collected since the last time that training was triggered.
+
+        """
+        num_samples_collected = sum(num_samples_collected.values())
+        self.curr_num_samples_collected += num_samples_collected
+        if self.curr_num_samples_collected > self.config["train_batch_size"]:
+            self.curr_num_samples_collected = 0
+            training_intensity = int(self.config["training_intensity"] or 1)
+            # for _ in range(training_intensity):
+            # rand_actor = random.choice(self.replay_actors)
+            replay_samples_ready: Dict[ActorHandle, T] = asynchronous_parallel_requests(
+                remote_requests_in_flight=self.remote_replay_requests_in_flight,
+                actors=self.replay_actors,
+                ray_wait_timeout_s=0.0,
+                num_requests_to_launch=training_intensity,
+                max_remote_requests_in_flight_per_actor=float("inf"),
+                remote_fn=lambda actor: actor.replay(),
+            )
+            for replay_actor, sample_batches in replay_samples_ready.items():
+                for sample_batch in sample_batches:
+                    self.replay_sample_batches.append((replay_actor, sample_batch))
+
+        replay_samples_ready: Dict[ActorHandle, T] = asynchronous_parallel_requests(
+            remote_requests_in_flight=self.remote_replay_requests_in_flight,
+            actors=self.replay_actors,
+            ray_wait_timeout_s=0.0,
+            max_remote_requests_in_flight_per_actor=float("inf"),
+            remote_fn=None,
+        )
+        for replay_actor, sample_batches in replay_samples_ready.items():
+            for sample_batch in sample_batches:
+                self.replay_sample_batches.append((replay_actor, sample_batch))
+
+        # add the sample batches to the learner queue
+        while self.replay_sample_batches:
+            try:
+                item = self.replay_sample_batches[0]
+                # the replay buffer returns none if it has not been filled to
+                # the minimum threshold yet.
+                if item:
+                    self.learner_thread.inqueue.put(
+                        self.replay_sample_batches[0], timeout=0.001
+                    )
+                    self.replay_sample_batches.pop(0)
+            except queue.Full:
+                break
+
+    def sample_from_replay_buffer_place_on_learner_queue_blocking(
```

Comment:
wait, why do we need both blocking and non-blocking version?

Reply:
ah we don't. I was using this for debugging, and I'll delete it after i'm done benchmarking this APEX implementation.